### PR TITLE
Smarter cleaning rules for packages with SemVer

### PR DESCRIPTION
**Pull request by ErikApption**
Smarter cleaning rules for packages

- MaxHistoryPerMajorVersion: Maximum number of major versions
- MaxHistoryPerMinorVersion: Maximum number of minor versions for each major version
- MaxHistoryPerPatch: Maximum number of patch versions for each major + minor version
- MaxHistoryPerPrerelease: Maximum number of prerelease versions for each major + minor + patch version and prerelease type. if you have `beta` and `alpha` this will keep `MaxHistoryPerPrerelease` versions for both `beta` and `alpha`.

### CODE REVIEW

File: src/BaGetter.Core/Configuration/BaGetterOptions.cs
Comment (robford123) [inline src/BaGetter.Core/Configuration/BaGetterOptions.cs:1]: Given the number of settings can we consider creating a new section in the config - maybe called "Retention" or something similar? That way it can be passed around as a single entity and will be easier to extend in the future if needed. To avoid breaking existing users we would need to retain the original MaxVersionsPerPackage setting and it's logic too or at minimum provide a configuration validation error/warning with a clear migration path.

Comment (ErikApption) [reply src/BaGetter.Core/Configuration/BaGetterOptions.cs:1]: Thanks @robford123 for looking into the PR. The `MaxVersionsPerPackage` hasn't gone yet into a release from what I can see so not sure why we would need a migration path. I like the idea of a retention section; I'll try to get these changes in the next couple days.

Comment (ErikApption) [reply src/BaGetter.Core/Configuration/BaGetterOptions.cs:1]: Updated the docs and added a "Retention" optional section

**Comment by ErikApption** [comment]
@robford123  Hey Rob, any chances you could look at integrating the PR?

File: src/BaGetter.Core/Configuration/BaGetterOptions.cs [hunk 1]
```diff
@@ -49,11 +49,7 @@ public class BaGetterOptions
     /// </summary>
     public uint MaxPackageSizeGiB { get; set; } = 8;
 
-    /// <summary>
-    /// If this is set to a value, it will limit the number of versions that can be pushed for a package.
-    /// the older versions will be deleted.
-    /// </summary>
-    public uint? MaxVersionsPerPackage { get; set; } = null;
```
Comment (seriouz) [inline src/BaGetter.Core/Configuration/BaGetterOptions.cs:56]: What about people, who are currently using this setting? You must set it deprecated/obsolete.
If they are using it currently you should fill the new `RetentionOptions` with the value of `MaxVersionsPerPackage`. A message in the console should point them to the problem so they can move to the new configuration.

File: src/BaGetter.Core/Indexing/PackageDeletionService.cs [hunk 2]
```diff
@@ -90,21 +91,106 @@ private async Task<bool> TryHardDeletePackageAsync(string id, NuGetVersion versi
         return found;
     }
 
-    public async Task<int> DeleteOldVersionsAsync(Package package, uint maxPackages, CancellationToken cancellationToken)
+    private List<NuGetVersion> GetValidVersions<S, T>(IEnumerable<NuGetVersion> versions, Func<NuGetVersion, S> getParent, Func<NuGetVersion,T> getSelector, int versionsToKeep)
```
[the remark this answers is not in the document]
Comment (seriouz) [reply src/BaGetter.Core/Indexing/PackageDeletionService.cs:94]: Make it to `IList`.

File: src/BaGetter.Core/Indexing/PackageDeletionService.cs [hunk 3]
```diff
@@ -90,21 +91,106 @@ private async Task<bool> TryHardDeletePackageAsync(string id, NuGetVersion versi
...
+                .Where(lb => lb is not null)
+                .Distinct();
+
+            var allPreReleaseValidVersions = new HashSet<NuGetVersion>();
+            foreach (var preReleaseType in prereleaseTypes)
+            {
+                var preReleaseVersions = preReleases.Where(p => p.ReleaseLabels!.FirstOrDefault() == preReleaseType
+                        && GetPreReleaseBuild(p) is not null).ToList();
+
+
```
Comment (seriouz) [inline src/BaGetter.Core/Indexing/PackageDeletionService.cs:153]:
```suggestion
```

File: src/BaGetter.Core/Indexing/PackageDeletionService.cs [hunk 4]
```diff
@@ -90,21 +91,106 @@ private async Task<bool> TryHardDeletePackageAsync(string id, NuGetVersion versi
...
             if (await TryHardDeletePackageAsync(package.Id, version.Version, cancellationToken)) deleted++;
         }
         return deleted;
     }
+
+    /// <summary>
+    /// If we have 1.1.1-alpha.1 , this will return 1
+    /// or null if not valid
+    /// </summary>
+    /// <returns></returns>
```
Comment (seriouz) [inline src/BaGetter.Core/Indexing/PackageDeletionService.cs:177]:
```suggestion
    /// <summary>
    /// Tries to get the version number of a pre-release build.<br/>
    /// If we have 1.1.1-alpha.1 , this will return 1 or <c>null</c> if not valid.
    /// </summary>
    /// <returns>The version as <c>int</c> or <c>null</c> if not found.</returns>
```

File: tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs [hunk 5]
```diff
@@ -151,9 +156,6 @@ public async Task IndexIMAsync_WhenPrereleasePackageAlreadyExists_AndOverwritePr
         var stream = new MemoryStream();
         builder.Save(stream);
 
-        _storage.Setup(s => s.DeleteAsync(builder.Id, builder.Version, default)).Returns(Task.CompletedTask);
-        _storage.Setup(s => s.SavePackageContentAsync(It.Is<Package>(p => p.Id == builder.Id && p.Version.ToString() == builder.Version.ToString()), stream, It.IsAny<FileStream>(), default, default, default)).Returns(Task.CompletedTask);
-
```
Comment (seriouz) [inline tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs:156]: Why did you remove the tests for Delete and Save?

File: tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs [hunk 6]
```diff
@@ -117,9 +125,6 @@ public async Task IndexIMAsync_WhenPackageAlreadyExists_AndOverwriteAllowed_Inde
         var stream = new MemoryStream();
         builder.Save(stream);
 
-        _storage.Setup(s => s.DeleteAsync(builder.Id, builder.Version, default)).Returns(Task.CompletedTask);
-        _storage.Setup(s => s.SavePackageContentAsync(It.Is<Package>(p => p.Id == builder.Id && p.Version.ToString() == builder.Version.ToString()), stream, It.IsAny<FileStream>(), default, default, default)).Returns(Task.CompletedTask);
```
Comment (seriouz) [inline tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs:121]: Why did you remove the tests for Delete and Save?

File: tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs [hunk 7]
```diff
@@ -185,9 +187,6 @@ public async Task IndexIMAsync_WhenPrereleasePackageAlreadyExists_AndOverwriteFo
...
```
Comment (seriouz) [inline tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs:189]: Why did you remove the tests for Delete and Save?

**Review by seriouz — COMMENTED** [review]
Some thing are on the todo before we can merge this.
Why did you remove so many lines in the tests? Better would be to create a complete new test file for the smart cleaning only.

File: src/BaGetter.Core/Configuration/BaGetterOptions.cs (hunk 1 above)
Comment (ErikApption) [reply src/BaGetter.Core/Configuration/BaGetterOptions.cs:56]: Agree - I was hoping this PR would be rolled in release before my other one that had the `MaxVersionsPerPackage`

File: tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs (hunk 5 above)
Comment (ErikApption) [reply tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs:156]: [PackageIndexingServiceInMemoryTests.cs](https://github.com/bagetter/BaGetter/blob/main/tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs) is a clone of [PackageIndexingServiceTests.cs](https://github.com/bagetter/BaGetter/blob/main/tests/BaGetter.Core.Tests/Services/PackageIndexingServiceTests.cs) that is there  to test the auto deletion. `_storage` uses an actual implementation instead of mock classes and assertions. The implementation provides an in-memory database and lets me test the whole workflow from addition to auto-deletion - at the same time this makes it more difficult to keep these assertions.

File: tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs (hunk 7 above)
Comment (ErikApption) [reply tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs:189]: same comment as above

File: tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs (hunk 6 above)
Comment (ErikApption) [reply tests/BaGetter.Core.Tests/Services/PackageIndexingServiceInMemoryTests.cs:121]: same as above

**Comment by ErikApption** [comment]
@seriouz I did address your feedback - not sure where github put all my comments, but 
- tests: the original tests are all in there with the assertions. I did duplicate the tests to run with full in-memory instances and state instead of stateless mockups
- rolled back `MaxVersionsPerPackage` parameter and added Obsolete and log error message

**Review by seriouz — APPROVED** [review]
Thanks for adding the requested changes.

**Comment by Regenhardt** [comment]
Ok just to make sure I'm not confused here:

 - `MaxHistory`: How many versions to keep?
 - `PerMajorVersion`: For each major version

So with e.g. `MaxHistoryPerMajorVersion: 5` this would mean, for major version 1 (i.e. versions 1.*.*), it would keep at most 5 versions, and when I start pushing versions with major version 2 (i.e. 2.*.*), the counter restarts and the newest 5 versions with major version 2 are gonna be kept?

**Comment by ErikApption** [comment]
> So with e.g. `MaxHistoryPerMajorVersion: 5` this would mean, for major version 1 (i.e. versions 1._._), it would keep at most 5 versions, and when I start pushing versions with major version 2 (i.e. 2._._), the counter restarts and the newest 5 versions with major version 2 are gonna be kept?

IIRC, pretty sure this means that once you read version `6._._` it will delete version 1._._ because it checks how many major versions are stored and deletes the older ones. I added the major version setting for the sake of completeness but the ones that do matter the most in my use case are the minor and patch limits.

**Comment by Regenhardt** [comment]
So `MaxHistoryPerMinorVersion: 5` means as soon as I push `1.5.0`, it will delete `1.0.0`? That is incorrect naming then IMO as "Per minor version" means "for one minor version" so "5 per minor version" means I can have 5 patch or prerelease versions for each one minor version. It should be `MaxMinorVersionsPerMajorVersion` or something.

**Comment by seriouz** [comment]
You are mostly right. If you push 1.6.0 the 1.0.0 will be delete and the in-between versions are kept.
He wrote in the docs what every config does, did you see it? I personally prefer the short naming more when the docs are understandable.

**Comment by Regenhardt** [comment]
Oh right, somehow managed to read past that.  
In that case, the naming isn't just confusing but actually wrong. The variable says "PerMajor", but then it's the number of major versions to be retained. 

As a short name it has to be ‘MaxMajorVersions‘ for how many major versions are retained.

**Comment by Regenhardt** [comment]
Or of course "MaxHistoryMajorVersions" if we don't want to change too much. The important part is that it doesn't say "per major" but "versions", plural and without per, unless we want to specify exactly per what the versions are retained (in case the next bigger version counter isn't obvious enough).

**Comment by seriouz** [comment]
If we rename the variables I am for these short ones you mentioned:

- MaxMajorVersions
- MaxMinorVersions
- MaxPatchVersions
- MaxPrepatchVersions
